agent/handlers/v2_server_setup.go

```diff
@@ -21,78 +21,87 @@ import (
 	"github.com/aws/amazon-ecs-agent/agent/config"
 	"github.com/aws/amazon-ecs-agent/agent/credentials"
 	"github.com/aws/amazon-ecs-agent/agent/engine/dockerstate"
+	handlersutils "github.com/aws/amazon-ecs-agent/agent/handlers/utils"
 	"github.com/aws/amazon-ecs-agent/agent/handlers/v1"
 	"github.com/aws/amazon-ecs-agent/agent/handlers/v2"
+	"github.com/aws/amazon-ecs-agent/agent/handlers/v3"
 	"github.com/aws/amazon-ecs-agent/agent/logger/audit"
-	"github.com/aws/amazon-ecs-agent/agent/logger/audit/request"
 	"github.com/aws/amazon-ecs-agent/agent/stats"
 	"github.com/aws/amazon-ecs-agent/agent/utils"
 	"github.com/cihub/seelog"
 	"github.com/didip/tollbooth"
+	"github.com/gorilla/mux"
 )
 
 const (
 	// readTimeout specifies the maximum duration before timing out read of the request.
 	// The value is set to 5 seconds as per AWS SDK defaults.
 	readTimeout = 5 * time.Second
+
 	// writeTimeout specifies the maximum duration before timing out write of the response.
 	// The value is set to 5 seconds as per AWS SDK defaults.
 	writeTimeout = 5 * time.Second
 )
 
-// v2SetupServer creates the HTTP server for serving task/container metadata,
-// task/container stats and IAM Role credentials for tasks,
 func v2SetupServer(credentialsManager credentials.Manager,
 	auditLogger audit.AuditLogger,
 	state dockerstate.TaskEngineState,
 	cluster string,
 	statsEngine stats.Engine,
 	steadyStateRate int,
 	burstRate int) *http.Server {
-	serverMux := http.NewServeMux()
-	// Credentials handlers
-	serverMux.HandleFunc(credentials.V1CredentialsPath,
+	muxRouter := mux.NewRouter()
+
+	// Set this so that for request like "/v3//metadata/task", the Agent will pass
+	// it to task metadata handler instead of returning a 301 error.
+	muxRouter.SkipClean(true)
```

Comment:
Is this a new feature or are we preserving existing behavior?

Reply:
This is specifically for the V3 handlers, it's used to handle the case where the v3 endpoint ID is empty in the path (path like `/v3//metadata/task`), it sounds reasonable for us to return 400 bad request instead of 301 for empty v3 endpoint ID, I discussed this with @richardpen.